Sources/_CryptoExtras/ChaCha20CTR/ChaCha20CTR.swift

```diff
@@ -0,0 +1,110 @@
+        ///   - message: The message to encrypt
+        ///   - key: A 256-bit encryption key
+        ///   - counter: A 4 byte counter (UInt32), defaults to 0
+        ///   - nonce: A 12 byte nonce for ChaCha20 encryption. The nonce must be unique for every use of the key to seal data.
+        /// - Returns: The encrypted ciphertext
+        /// - Throws: CipherError errors
+        /// - Warning: You most likely want to use the ChaChaPoly implemention with AuthenticatedData available at `Crypto.ChaChaPoly`
+        public static func encrypt<Plaintext: DataProtocol>
+        (_ message: Plaintext, using key: SymmetricKey, counter: Insecure.ChaCha20CTR.Counter = Counter(), nonce: Insecure.ChaCha20CTR.Nonce) throws -> Data {
+            return try ChaCha20CTRImpl.encrypt(key: key, message: message, counter: counter.counter, nonce: nonce.bytes)
+        }
+    }
+}
+
+extension Insecure.ChaCha20CTR {
+    public struct Nonce: ContiguousBytes, Sequence {
+        let bytes: Data
+
+        /// Generates a fresh random Nonce. Unless required by a specification to provide a specific Nonce, this is the recommended initializer.
+        public init() {
+            var data = Data(repeating: 0, count: Insecure.ChaCha20CTR.nonceByteCount)
+            data.withUnsafeMutableBytes {
+                assert($0.count == Insecure.ChaCha20CTR.nonceByteCount)
+                $0.initializeWithRandomBytes(count: Insecure.ChaCha20CTR.nonceByteCount)
+            }
+            self.bytes = data
+        }
+
+        public init<D: DataProtocol>(data: D) throws {
+            if data.count != Insecure.ChaCha20CTR.nonceByteCount {
+                throw CryptoKitError.incorrectParameterSize
+            }
+
+            self.bytes = Data(data)
+        }
+
+        public func withUnsafeBytes<R>(_ body: (UnsafeRawBufferPointer) throws -> R) rethrows -> R {
+            return try self.bytes.withUnsafeBytes(body)
+        }
+
+        public func makeIterator() -> Array<UInt8>.Iterator {
+            self.withUnsafeBytes({ buffPtr in
+                Array(buffPtr).makeIterator()
+            })
+        }
+    }
+
+    public struct Counter: ContiguousBytes {
+        let counter: UInt32
+
+        /// Generates a fresh Counter set to 0. Unless required by a specification to provide a specific Counter, this is the recommended initializer.
+        public init() {
+            self.counter = 0
+        }
+
+        /// Explicitly set the Counter's offset using a little endian byte sequence
+        public init<D: DataProtocol>(data: D) throws {
+            if data.count != Insecure.ChaCha20CTR.counterByteCount {
+                throw CryptoKitError.incorrectParameterSize
+            }
+
+            if data.regions.count == 1 {
+                self.counter = data.regions.first!.withUnsafeBytes { $0.load(as: UInt32.self) }
+            } else {
+                self.counter = Array(data).withUnsafeBytes { $0.load(as: UInt32.self) }
+            }
```

Comment:
Ok, we can avoid some unsafe code here.

All `DataProtocol` are also `Collection<UInt8>`, so we can safely load this using a safe method. Additionally, it'll solve the endianness problem that this code has, where it claims to use little-endian but actually uses native-endian.

Let's replace it with:

```swift
if data.count != Insecure.ChaCha20CTR.counterByteCount {
    throw CryptoKitError.incorrectParameterSize
}

let startIndex = data.startIndex
self.counter = (
    (UInt32(data[data.index(startIndex, offsetBy: 0)]) << 0) |
    (UInt32(data[data.index(startIndex, offsetBy: 1)]) << 8) |
    (UInt32(data[data.index(startIndex, offsetBy: 2)]) << 16) |
    (UInt32(data[data.index(startIndex, offsetBy: 3)]) << 24) |
)
```

Reply:
Nice! 👍